### PR TITLE
Indirect Resolver for Constant Values

### PR DESCRIPTION
**Pull request by mahaloz**
* Added a new resolver to default resolvers
* Resolver handles jmps/calls to underlying constant values
* Constants found through forward propagation to the target jump

### CODE REVIEW

**Comment by mahaloz** [comment]
Working on these bugs still.

File: angr/analyses/cfg/indirect_jump_resolvers/const_resolver.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
+import pyvex
+from .resolver import IndirectJumpResolver
+from ...code_location import CodeLocation
+from ...propagator import vex_vars
+import logging
```
Comment (ltfish) [inline angr/analyses/cfg/indirect_jump_resolvers/const_resolver.py:5]: Please rearrange these imports following the following order:

```
Python packages

3rd party packages

other angr packages

indirect imports inside angr
```

File: angr/analyses/cfg/indirect_jump_resolvers/const_resolver.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+
+class ConstantResolver(IndirectJumpResolver):
+    def __init__(self, project):
+        super().__init__(project, timeless=True)
+
+    def filter(self, cfg, addr, func_addr, block, jumpkind):
+        """
+        This function acts as a filter for possible instructions.
+
```
Comment (ltfish) [inline angr/analyses/cfg/indirect_jump_resolvers/const_resolver.py:17]: Useless blank line.

File: angr/analyses/cfg/indirect_jump_resolvers/const_resolver.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        # we support both an indirect call and jump since the value can be resolved
+        if jumpkind in ('Ijk_Boring', 'Ijk_Call'):
+            return True
+
+        return False
+
+    def resolve(self, cfg, addr, func_addr, block, jumpkind):
+        """
+        This function does the actual resolve. Our process is easy:
+
```
Comment (ltfish) [inline angr/analyses/cfg/indirect_jump_resolvers/const_resolver.py:29]: Redundant blank line.

File: angr/analyses/propagator/engine_vex.py [hunk 4]
```diff
@@ -29,7 +29,9 @@ def _process(self, state, successors, block=None, whitelist=None, **kwargs):  #
                 # pop ret from the stack
                 sp_offset = self.arch.sp_offset
                 sp_value = state.load_register(sp_offset, self.arch.bytes)
-                state.store_register(sp_offset, self.arch.bytes, sp_value + self.arch.bytes)
+                #XXX: unsure if this is ok since now making a call without a ret
```
Comment (ltfish) [inline angr/analyses/propagator/engine_vex.py:32]: What do you mean?

File: angr/analyses/propagator/engine_vex.py [hunk 5]
```diff
@@ -29,7 +29,9 @@ def _process(self, state, successors, block=None, whitelist=None, **kwargs):  #
...
+                if sp_value:
```
Comment (ltfish) [inline angr/analyses/propagator/engine_vex.py:33]: I suppose you want to check `if sp_value is not None`. Make it explicit.

**Comment by mahaloz** [comment]
All checks, except for API check work. Need a merge over in angr-docs also.

**Comment by ltfish** [comment]
An optimization: If a function has multiple indirect jumps, are we going to perform multiple full-function propagation runs?

File: angr/analyses/cfg/indirect_jump_resolvers/default_resolvers.py [hunk 6]
```diff
@@ -23,6 +23,7 @@
     'MIPS64': {
         cle.MetaELF: [ MipsElfFastResolver, ],
     },
+
```
Comment (ltfish) [inline angr/analyses/cfg/indirect_jump_resolvers/default_resolvers.py:26]: Remove this extra new line please.

File: angr/analyses/cfg/indirect_jump_resolvers/default_resolvers.py [hunk 7]
```diff
@@ -32,10 +33,9 @@
     'ARMCortexM': {
         cle.MetaELF: [ ArmElfFastResolver, ]
     },
-    'ALL': [ JumpTableResolver ],
+    'ALL': [ JumpTableResolver, ConstantResolver ],
```
Comment (ltfish) [inline angr/analyses/cfg/indirect_jump_resolvers/default_resolvers.py:36]: Shall we swap them and run `ConstantResolver` first?

File: angr/analyses/cfg/indirect_jump_resolvers/default_resolvers.py [hunk 8]
```diff
@@ -32,10 +33,9 @@
...
 }
 
-
```
Comment (ltfish) [inline angr/analyses/cfg/indirect_jump_resolvers/default_resolvers.py:38]: Don't remove this new line.

File: angr/knowledge_plugins/propagations.py [hunk 9]
```diff
@@ -0,0 +1,53 @@
+from .plugin import KnowledgeBasePlugin
+
+
+class Propagations(KnowledgeBasePlugin):
```
Comment (ltfish) [inline angr/knowledge_plugins/propagations.py:4]: I don't see where this new KB plugin is used. Is that because this PR is still WIP?

Comment (mahaloz) [reply angr/knowledge_plugins/propagations.py:4]: Added WIP label

**Comment by mahaloz** [comment]
Optimizing is not currently possible on a function level, since propagations are called and created while we do not have a full CFG -- thus we can't propagate across a full function since it is not complete. Added a KB plugin to allow to do this in the future when we either use it for second pass indirect jump resolving, or for various other re-uses of propagations.

**Pull request merged**